### PR TITLE
Data types coverage & bug corrections

### PR DESCRIPTION
**Pull request by AntoninPoche**
# 1. Data type coverage

The first part of this pull request concerns Xplique data type coverage extension. It first modifies some methods to extend the coverage, then tests the coverage and shows the application in a tutorial.

## 1.1 Non-square images

`SobolAttributionMethod` and `HsicAttributionImage` did not support non-square images due to the use of `cv2.resize()` which permutes the h and w dimensions. It was thus changed to `tf.image.resize()` in b77c5f3 and tested in 8b8465f.

## 1.2 Time series & Tabular data

Xplique was initially designed for images but it also supports attribution methods for tabular data and now time series data.

Xplique conciders data with:
- 4 dimensions as images.
- 3 dimensions as time series.
- 2 dimensions as tabular data.

### New attribution time series tutorial [Open In Colab](https://colab.research.google.com/drive/1h0lThbcP5d2VKtRxwLG8z7KC8PExcVIA)

To show how to use Xplique on time series a new tutorial was designed: [Attributions: Time Series and Regression](https://colab.research.google.com/drive/1h0lThbcP5d2VKtRxwLG8z7KC8PExcVIA), the link was added to the documentation in 6855221.

### Time series attribution plot

Furthermore, for this tutorial, time series attribution plots were updated in 397c681. The new name is `xplique.plots.plot_timeseries_attributions` with an API very similar to `xplique.plots.plot_attributions`. Here is an example from the tutorial on temperature forecasting for the next 24 hours based on weather data from the last 48 hours.

image

### Update some methods

`Rise` method is now applicable to tabular and time series thanks to f1c7b17, it was then tested in 8b8465f. The main modifications were for the design of the perturbation masks where a switch case on the data types was made.

Then, `Lime`, and `Kernelshap` did not support time series natively, it was modified in 5d09b46 and tested in 8b8465f. This could be summarized as adding a default `map_to_interpret_space` for time series

### Relative documentation

To clarify which method supports tabular data and time series, the documentation was updated and the table of attribution methods was simplified in c131b7d:

| **Attribution Method** | Type of Model            | Images         | Time Series and Tabular Data |
| :--------------------- | :----------------------- | :------------: | :--------------------------: |
| Deconvolution          | TF                       | C✔️ OD❌ SS❌ | C✔️ R✔️ | 
| Grad-CAM               | TF                       | C✔️ OD❌ SS❌ | ❌      |
| Grad-CAM++             | TF                       | C✔️ OD❌ SS❌ | ❌      |
| Gradient Input         | TF, PyTorch**            | C✔️ OD✔️ SS✔️ | C✔️ R✔️ | 
| Guided Backprop        | TF                       | C✔️ OD❌ SS❌ | C✔️ R✔️ | 
| Integrated Gradients   | TF, PyTorch**            | C✔️ OD✔️ SS✔️ | C✔️ R✔️ | 
| Kernel SHAP            | TF, PyTorch**, Callable* | C✔️ OD✔️ SS✔️ | C✔️ R✔️ |
| Lime                   | TF, PyTorch**, Callable* | C✔️ OD✔️ SS✔️ | C✔️ R✔️ |
| Occlusion              | TF, PyTorch**, Callable* | C✔️ OD✔️ SS✔️ | C✔️ R✔️ | 
| Rise                   | TF, PyTorch**, Callable* | C✔️ OD✔️ SS✔️ | C✔️ R✔️ | 
| Saliency               | TF, PyTorch**            | C✔️ OD✔️ SS✔️ | C✔️ R✔️ | 
| SmoothGrad             | TF, PyTorch**            | C✔️ OD✔️ SS✔️ | C✔️ R✔️ |
| SquareGrad             | TF, PyTorch**            | C✔️ OD✔️ SS✔️ | C✔️ R✔️ |
| VarGrad                | TF, PyTorch**            | C✔️ OD✔️ SS✔️ | C✔️ R✔️ |
| Sobol Attribution      | TF, PyTorch**            | C✔️ OD✔️ SS✔️ | 🔵      |
| Hsic Attribution       | TF, PyTorch**            | C✔️ OD✔️ SS✔️ | 🔵      |
| FORGrad enhancement    | TF, PyTorch**            | C✔️ OD✔️ SS✔️ | ❌      | 

TF : Tensorflow compatible
C : [Classification](https://deel-ai.github.io/xplique/latest/api/attributions/classification/) | R : [Regression](https://deel-ai.github.io/xplique/latest/api/attributions/regression/) |
OD : [Object Detection](https://deel-ai.github.io/xplique/latest/api/attributions/object_detection/) | SS :  [Semantic Segmentation (SS)](https://deel-ai.github.io/xplique/latest/api/attributions/semantic_segmentation/)

\* : See the [Callable documentation](https://deel-ai.github.io/xplique/latest/api/attributions/callable/)

** : See the [Xplique for PyTorch documentation](https://deel-ai.github.io/xplique/latest/api/attributions/pytorch/), and the [**PyTorch models**: Getting started](https://colab.research.google.com/drive/1bMlO29_0K3YnTQBbbyKQyRfo8YjvDbhe) notebook.

✔️ : Supported by Xplique | ❌ : Not applicable | 🔵 : Work in Progress

### Test metrics

To complete time series integration in Xplique, a test to ensure metrics support them was added in fa356cc. In fact, `MuFidelity` was modified to support time series and tabular data in ec0fb20.

## 1.3 Image explanation shape harmonization

For image explanation, depending on the method, the explanation shape could be either $(n, h, w)$, $(n, h, w, 1)$, or $(n, h, w, 3)$. It was decided to harmonize it to $(n, h, w, 1)$. This was modified in 0ae74de, tested in 8b8465f, and documented in 7a95dc7. Furthermore, unit tests needed to be modified as they hard-coded the expected output shape. Thus they were modified in 1f281c7.

### Reducer for gradient-based methods

For images, most gradient-based provide a value for each channel, however, for consistency, it was decided that for images, explanations will have the shape $(n, h, w, 1)$. Therefore, gradient-based methods need to reduce the channel dimension of their image explanations and the `reducer` parameter chooses how to do it among {`"mean"`, `"min"`, `"max"`, `"sum"`, `None`}. In the case `None` is given, the channel dimension is not reduced. The default value is `"mean"` for methods except `Saliency` which is `"max"` to comply with the paper and `GradCAM` and `GradCAMPP` which are not concerned. 

---

# 2. Bugs correction

The second part of this pull request is to solve pending issues.

## 2.1 Memories problem

Indeed, among the reported issues several concerned memory management.

### SmoothGrad, VarGrad, and SquareGrad  issue #137

`SmoothGrad`, `VarGrad`, and `SquareGrad` created tensors of shape `(nb_samples, *inputs.shape[1:])` and ignored the `batch_size` parameter. Now all methods inherit from `GradientStatistic` and use an implement method to initialize, update, and get an online statistic, either the mean, the mean of squares, or the variance depending on the method. Computing those statistics online allow the method to make inference batch by batch and take `batch_size` into account.

Furthermore, when `batch_size` is greater than several times `nb_samples`, several inputs are treated in the same batch. Thus allowing a greater adaptation to the memory capacity. This was implemented in 7076e9c and 4e54fdf, then tested in df4f819.

### MuFidelity issue #137

The metric MuFidelity had the same problem as the three previous methods, creating a tensor too large for the memory when passed to the model. It was corrected in the same way in ec0fb20 and tested in 5723855.

### HsicAttributionMethod

This method had a different memory problem the `batch_size` for the model was used correctly, however, when computing the estimator a tensor of size `grid_size**2 * nb_design**2` was created. However, for big images and/or small objects in images, the `grid_size` needs to be increased, furthermore, for the estimator to converge, `nb_design` should also be increased accordingly. Which creates out-of-memory errors.

Thus an `estimator_batch_size` (different from the initial `batch_size`) was introduced in 31a9674 to batch over the `grid_size**2` dimension, and tested in 3471882. The default value is `None`, thus conserving the default behavior of the method, but when an out-of-memory occurs, setting an `estimator_batch_size` smaller than `grid_size**2` will reduce the memory cost of the method.

### Metrics input types issues #102 and #128 

Errors were reported when using other inputs than `np.ndarray`, the problem was that `self.inputs` initialized and sanitized in `BaseAttributionMetric.__init__()` were overwritten in the `__init__()` other the different metrics. It was corrected in a3a76bf by making sure `inputs` used go through `numpy_sanitize`, then it was tested in daba2b4.

### Feature visualization latent dtype issue #131 

In issue #131, @RuoyuChen10 reported a bug in the feature visualization module. A conflict in dtype between the model internal dtype and Xplique dtype. We made sure that the dtype used for the conflicting computation was the model's internal dtype. It was implemented in c3ff31b.

## 2.3 Other corrections

Naturally, other problems were reported to us outside of issues or discovered by the team, we also addressed these.

### Some refactorization

`Lime`:
Pylint was not happy with all the if-else branches to determine default values for `map_to_interpret_space` and `ref_value`. Those cases were moved from `explain()` to `_set_shape_dependant_parameters()`, the latter being called by `explain()`. This was done to remove the `_compute()` method and place the main part of the code in the `explain()` method for consistency between methods. This was done in 1e3c993.

### Typo and small fixes

In `HsicAttributionMethod` and `SobolAttributionMethod` there was a difference between the documentation of the `perturbation_function` and the actual code. It was corrected in abd7975.

Other typos and small mistakes were corrected in fb58c3a and 59d441a.

For Craft, there were some remaining prints, but they may be useful, thus Craft's methods with print now take a `verbose` parameter. Implemented in d696f93.

In Craft tests, a PyTorch tensor conversion bug was resolved in d6e1465.

Pylint raised `no-member` errors for some of PyTorch's functions, this comes from the fact that Pylint does not recognize dynamically set members. Therefore, `no-member` was added to the list of disabled Pylint errors in 0dcc91c.

### CODE REVIEW

**Review by Agustin-Picard — COMMENTED** [review]
Great work Antonin! 🔥 I just left some minor comments ;)

File: xplique/plots/image.py [hunk 1]
```diff
@@ -58,6 +58,20 @@ def _clip_percentile(tensor: Union[tf.Tensor, np.ndarray],
     return tensor
 
 
+def _clean_explanation(explanation: Union[tf.Tensor, np.ndarray],
```
[the remark this answers is not in the document]
Comment (AntoninPoche) [reply xplique/plots/image.py:61]: Done

File: tests/metrics/test_fidelity.py [hunk 2]
```diff
@@ -1,18 +1,18 @@
 import tensorflow as tf
 import numpy as np
 
-from ..utils import (generate_model, generate_regression_model, generate_timeseries_model, 
+from tests.utils import (generate_model, generate_regression_model, generate_timeseries_model,
```
[the remark this answers is not in the document]
Comment (AntoninPoche) [reply tests/metrics/test_fidelity.py:4]: It was for debugging, I change it back.

**Review by Agustin-Picard — APPROVED** [review]
Awesome work, LGTM!
